common/kernel/timing.cc

```diff
@@ -1350,429 +920,181 @@ CriticalPath build_critical_path_report(Context *ctx, ClockPair &clocks, const P
+        return false;
+    };
 
-        // Clock to clock delays for xpaths
-        dict<ClockPair, delay_t> xclock_delays;
-        for (auto &report : xclock_reports) {
-            const auto &clock1_name = report.clock_pair.start.clock;
-            const auto &clock2_name = report.clock_pair.end.clock;
+    std::sort(xclock_reports.begin(), xclock_reports.end(), cmp_crit_path);
 
-            const auto key = std::make_pair(clock1_name, clock2_name);
-            if (clock_delays.count(key)) {
-                xclock_delays[report.clock_pair] = clock_delays.at(key);
-            }
-        }
-
-        unsigned max_width_xca = 0;
-        unsigned max_width_xcb = 0;
-        for (auto &report : xclock_reports) {
-            max_width_xca = std::max((unsigned)format_event(report.clock_pair.start).length(), max_width_xca);
-            max_width_xcb = std::max((unsigned)format_event(report.clock_pair.end).length(), max_width_xcb);
-        }
+    clock_delays_ctx = clock_delays;
+}
 
-        // Check and report xpath delays for related clocks
-        if (!xclock_reports.empty()) {
-            for (auto &report : xclock_reports) {
-                const auto &clock_a = report.clock_pair.start.clock;
-                const auto &clock_b = report.clock_pair.end.clock;
+void TimingAnalyser::build_slack_histogram_report()
+{
+    auto &slack_histogram = result.slack_histogram;
 
-                const auto key = std::make_pair(clock_a, clock_b);
-                if (!clock_delays.count(key)) {
-                    continue;
-                }
+    for (domain_id_t dom_id = 0; dom_id < domain_id_t(domains.size()); ++dom_id) {
+        for (auto &ep : domains.at(dom_id).endpoints) {
+            auto &pd = ports.at(ep.first);
 
-                delay_t path_delay = 0;
-                for (const auto &segment : report.segments) {
-                    path_delay += segment.delay;
-                }
+            for (auto &req : pd.required) {
+                auto &capture = domains.at(req.first).key;
+                for (auto &arr : pd.arrival) {
+                    auto &launch = domains.at(arr.first).key;
 
-                // Compensate path delay for clock-to-clock delay. If the
-                // result is negative then only the latter matters. Otherwise
-                // the compensated path delay is taken.
-                auto clock_delay = clock_delays.at(key);
-                path_delay -= clock_delay;
-
-                float fmax = std::numeric_limits<float>::infinity();
-                if (path_delay < 0) {
-                    fmax = 1e3f / ctx->getDelayNS(clock_delay);
-                } else if (path_delay > 0) {
-                    fmax = 1e3f / ctx->getDelayNS(path_delay);
-                }
+                    // @gatecat: old timing analysis includes async path
+                    // The slack doesn't make much sense so I filter it out.
+                    if (launch.clock != capture.clock || launch.is_async())
```

Comment:
See the comment.

Reply:
Yes, agreed, looks like a good change to me